app/controllers/user_controller.php

```diff
@@ -2,37 +2,239 @@
+        }
+        if (self::MAX_POST_SIZE < $_SERVER['CONTENT_LENGTH']) {
+            $error = "Error! File Size too big! Can only upload 2 Mega Bytes";
+        }
+        if ($_FILES['avatar']['error'] === 4 && !isset($error)) {
+            $error = "No File was Selected";
+        }
+        if ($_FILES['avatar']['error'] > 0 && !isset($error)) {
+            $error = "Upload Error Occured";
+        }
+        if (isset($error)) {
+            $this->set(get_defined_vars());
+            return;
+        }
+        $file = $_FILES['avatar'];
+        $upload = new UploadImg($file);
+        if (!$upload->isFileAccepted()) {
+            $error = "File Upload Error";
+            $this->set(get_defined_vars());
+            return;
+        }
+        $upload->rename($user->username);
+        if (strpos($user->user_details['avatar'], 'default') === false) {
+            unlink(APP_DIR.'webroot'.$user->user_details['avatar']);
+        }
+        $saved = $upload->save('webroot/public_images/');
+        if (!$saved) {
+            $error = "Unexpected Error Occured";
+            $this->set(get_defined_vars());
+            return;
+        }
+        $user->saveAvatar($saved);
+        chmod(APP_DIR.'webroot'.$saved, 0755);
+        $this->set(get_defined_vars());
+    }
+
+    public function search()
+    {
+        $type = Param::get('type', false);
+        $query = Param::get('query', false);
+        if (!$type || !$query) {
+            redirect(url('user/profile'));
+        }
+        $user = new User($_SESSION['username']);
+        switch ($type) {
+            case ('user'):
+                $render_page = '/user/search';
+                $varname = 'query_results';
+                break;
+            case ('thread'):
+                $render_page = '/thread/index';
+                $varname = 'threads';
+                break;
+            case ('comment'):
+                
+                $render_page = '/comment/view';
+                $varname = 'comments';
+                break;
+            default:
+                redirect(url('user/profile'));
+                break;
+        }
+        $title = sprintf("Search for '%s' in %s", $query, $type);
+        $url_params = "?type={$type}&query={$query}";
+        $query_results = $user->search($type, $query);
+        $page = Param::get('page', 1);
+        $page = (int) $page;
+        $per_page = 5;
+        $pagination = new SimplePagination($page, $per_page);
+        $results = array();
+        for ($i = ($pagination->start_index - 1); $i < (($pagination->count)*$page); $i++) {
+            if (isset($query_results[$i])) {
+                $results[] = $query_results[$i];
+            }
         }
-        $home = '/user/index';
-        $user->username = $_SESSION['username'];
-        $user_id = Param::get('user_id',0);
-        $user_details = $user->getUserDetail($user_id);
+        $total = count($query_results);
+        ${$varname} = $results;
```

Comment:
but this is not a good way to pass variable and value from controller to view. it will be created in runtime and this may also be hard to debug or trace for others. may be you can just use this?

```php
$this->set('query_results', $results);
$this->set('threads', $results);
$this->set('comments', $results);
```